Add more granular control around the stash user and group.

Z3r0Sum opened the pull request:
- Give individuals the capability to manage the stash user and group
  outside the context of the stash module as other supporting
  directories or items might also be using these same resources.
- Given the above, individuals would still want the ability to manage
  these resources with puppet.  These changes support that functionality.
- Default values for the user_manage and group_manage parameters are set to
  true, so that the module maintains prior behavior/backwards compatibility.

Z3r0Sum committed:
Add more granular control around the stash user and group.

  - Give individuals the capability to manage the stash user and group
    outside the context of the stash module as other supporting
    directories or items might also be using these same resources.
  - Given the above, individuals would still want the ability to manage
    these resources with puppet.  These changes support that functionality.
  - Default values for the user_manage and group_manage parameters are set to
    true, so that the module maintains prior behavior/backwards compatibility.

Z3r0Sum committed:
Add unit tests for user and group management feature.

  - Add rspec-puppet unit tests to stash_install_deploy_spec and
    stash_install_staging_spec.

Correct test cases for user and group management feature.

  - The tests for the lack of user and group within the catalog was
    not properly being done when the group_manage and user_manage were
    set to false.
  - Add tests to satisfy the above requirements.

igalic commented on manifests/install.pp:
```diff
@@ -52,21 +54,25 @@
+      password_max_age => '99999',
+      managehome       => true,
+      uid              => $uid,
+      gid              => $gid,
```
what if someone decides to manage one, but not the other?

---

i realise this is stupid speculation, but if you give people the opportunity (to mess up), they **_will**_ use it!

Z3r0Sum replied:
@igalic  Not sure I follow.  Are you saying something like the following scenario:
- Manage user inside the module
- Do not manage group inside the module
- User forgets to manage group or fails to handle ordering.

If so, maybe I could check for the GID somehow and fail if I don't see it?  I don't see an issue potentially surfacing if they managed the group in the module and the user outside the module with the current methodology.

nibalizer replied:
I think you can do something like:

``` puppet
if $user_manage != $group_manage {
  fail("managing only one of user and group isn't supported by the module, use a wrapper class")
}
```

igalic replied:
yup

Z3r0Sum replied:
@nibalizer Awesome - thanks!  I'll try and get this updated later tonight.

Z3r0Sum committed:
Adjust user and group management scope.

  - In order to avoid potential dependency issues, this module will
    either manage both the stash user and the group or neither.
  - Cleanup tests - boolean attributes were not being expressed
    correctly and remove a mistakenly commented code block.

mkrakowitzer commented on manifests/install.pp:
```diff
@@ -52,21 +54,32 @@
-    managehome       => true,
-    uid              => $uid,
-    gid              => $gid,
+  #Only manager user and group together or not at all
```
I dont understand why we have two params for this? If we can not manage them independently then it should be the same parameter?

Z3r0Sum replied:
True. I'll refactor it.

mkrakowitzer commented on spec/classes/stash_install_deploy_spec.rb:
```diff
@@ -31,6 +31,76 @@
+          context 'when no user or group are specified' do
+            it do
+              should_not contain_user('stash').with({
+                'comment'          => 'Stash daemon account',
```
No need to set the user params here. Either the user exists or it doesn't.

mkrakowitzer commented on spec/classes/stash_install_deploy_spec.rb:
```diff
@@ -31,6 +31,76 @@
+            end 
+            it do
+              should_not contain_group('stash').with({
+                'ensure' => 'present',
```
No need to set params for groups here either.

Z3r0Sum replied:
@mkrakowitzer I had issues with that technique as it still found 'Group[stash]' metaparameters in the catalog, so the tests would fail:

Failure/Error: it {  should_not contain_group('stash') }
       expected that the catalogue would not contain Group[stash]

Is there a way around that?  I would like it to be cleaner if possible.

mkrakowitzer replied:
Its because you are passing a string as a param and not a bool

Change:
-            :manage_usr_grp  => 'true',
-            :manage_usr_grp  => true,
  and
-            :manage_usr_grp  => 'false',
-            :manage_usr_grp  => false,

See http://docs.puppetlabs.com/puppet/3.7/reference/lang_datatypes.html#automatic-conversion-to-boolean

You may want to test for this or always convert $manage_usr_grp to a bool with a function from stdlib such as str2bool.

Z3r0Sum replied:
Thanks for all the help. I'll fix this some point today.

Z3r0Sum committed:
Update the README

  - New parameter 'manage_usr_grp' explanation.

nibalizer commented on spec/classes/stash_install_staging_spec.rb:
```diff
@@ -32,6 +32,42 @@
+              :user  => 'mystashuser',
+              :group => 'mystashgroup'
+            }}
+            it {  should_not contain_user('stash') }
```
Is this what you meant to test here? I can see it being valid but seems like we should also be ensuring that user{'mystashuser'} is not managed as well.

Z3r0Sum replied:
Nice catch - I'll fix it.

Z3r0Sum committed:
Fix test data for rspec tests related to installation.

  - Fix the incorrect user and group names in the rspec examples.
